### PR TITLE
[SPARK-14462][ML][MLLIB] add the mllib-local build to maven pom

### PR DESCRIPTION
**Pull request by dbtsai**
## What changes were proposed in this pull request?

In order to separate the linear algebra, and vector matrix classes into a standalone jar, we need to setup the build first. This PR will create a new jar called mllib-local with minimal dependencies. The test scope will still depend on spark-core and spark-core-test in order to use the common utilities, but the runtime will avoid any platform dependency. Couple platform independent classes will be moved to this package to demonstrate how this work.
## How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by dbtsai** [comment]
+cc @mengxr @jkbradley @srowen @holdenk @MLnick

**Comment by dbtsai** [comment]
+cc @JoshRosen who may be able to give me insight on the MiMa failure caused by adding new jar.

**Comment by holdenk** [comment]
Some questions from a first pass:
Why did we duplicate `package-info.java` and `package.scala` from ml?
Could we move `SparkException` to common/sketch (or similar) so we don't need to change the exception type everywhere?

**Comment by JoshRosen** [comment]
To fix MiMa, you need to add a line to the `mimaProjects` definition to temporarily exclude the new project: https://github.com/dbtsai/spark/blob/b9870d6c62d75cd59698bd72751bc4f7854cd2e3/project/SparkBuild.scala#L254

Also, add a TODO and followup task for post-2.0-release to remove the exclude once 2.0 has been published and there is a previous artifact for MiMa to compare against.

**Comment by dbtsai** [comment]
@JoshRosen Thanks. It's working now.

@holdenk I thought each jar needs to have its own `package-info.java` to generate the Java doc and Scala doc. I'm now wondering if this will cause class conflict, and should it be that each package only have one `package-info.java`? We can move `SparkException` into `sketch` if this will be a stable platform independent project. Although I think we should use `IllegalArgumentException` here instead of `SparkException`. But I'm going to change back to have minimal changes in this PR.

**Comment by holdenk** [comment]
@dbtsai  Did you generate docs for this PR? I'm not sure how docs handles duplicate package classes between the components (might be ok but just something worth making sure).

**Comment by dbtsai** [comment]
test this please

File: core/pom.xml [hunk 1]
```diff
@@ -35,6 +35,11 @@
   <url>http://spark.apache.org/</url>
   <dependencies>
     <dependency>
+      <groupId>org.apache.spark</groupId>
```
Comment (srowen) [inline core/pom.xml]: This seems unrelated; core doesn't change right?

File: mllib-local/pom.xml [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    <sbt.project.name>mllib-local</sbt.project.name>
+  </properties>
+  <packaging>jar</packaging>
+  <name>Spark Project ML Local Library</name>
+  <url>http://spark.apache.org/</url>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.spark</groupId>
+      <artifactId>spark-core_${scala.binary.version}</artifactId>
```
Comment (srowen) [inline mllib-local/pom.xml]: Hm, can mllib-local tests actually need spark core? I can see needing core _test_ code in its tests (next entry)

File: dev/sparktestsupport/modules.py [hunk 3]
```diff
@@ -256,9 +256,21 @@ def __hash__(self):
 )
 
 
+mllib_local = Module(
+    name="mllib-local",
+    dependencies=[sketch],
```
Comment (mengxr) [inline dev/sparktestsupport/modules.py]: What classes do we need from `sketch`? It seems unrelated.

**Comment by mengxr** [comment]
@dbtsai Let us separate the project/build setup from the content. In this PR, we only need one dummy source file under `mllib-local` with a dummy test to ensure project setup and build. We can work on the content in a separate PR.

File: dev/sparktestsupport/modules.py (hunk 3 above)
Comment (dbtsai) [inline dev/sparktestsupport/modules.py]: I moved the `SparkExpcetion` into `sketch` so we can use it in mllib-local. If we decide not to use Saprk specific exception or annotation in `mllib-local`, I can remove it.

Comment (srowen) [inline dev/sparktestsupport/modules.py]: It doesn't belong in either sketch or mllib-local really. Since the purpose is to dissociate from spark-core, you should probably make a new exception type or examine whether generic exception types can be used instead.

File: core/pom.xml (hunk 1 above)
Comment (dbtsai) [inline core/pom.xml]: I moved the `SparkException` from `core` to `sketch` so I can use `SparkException` in `mllib-local`.

File: dev/sparktestsupport/modules.py (hunk 3 above)
Comment (dbtsai) [inline dev/sparktestsupport/modules.py]: Sounds fair! Actually, I think it's better to just use `IllegalArgumentException` here.

**Comment by dbtsai** [comment]
This is the minimal change for creating a new jar build. Let's wait the result of Jenkins. We'll move the code in a separate PR once this is merged. Thanks.

**Comment by dbtsai** [comment]
test this please

File: mllib/pom.xml [hunk 4]
```diff
@@ -63,25 +63,20 @@
       <version>${project.version}</version>
     </dependency>
     <dependency>
-      <groupId>org.scalanlp</groupId>
```
Comment (mengxr) [inline mllib/pom.xml]: Please keep the breeze dependency in MLlib's pom because we call its APIs directly. We can move versions to a common place.

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

**Comment by yhuai** [comment]
Looks like it breaks the build?

**Comment by mengxr** [comment]
Reverted. https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Compile/job/spark-master-compile-maven-hadoop-2.7/950/consoleFull

~~~
[info] Compiling 1 Scala source to /home/jenkins/workspace/spark-master-compile-maven-hadoop-2.7/mllib-local/target/scala-2.11/test-classes...
[error] missing or invalid dependency detected while loading class file 'SparkFunSuite.class'.
[error] Could not access type Logging in package org.apache.spark.internal,
[error] because it (or its dependencies) are missing. Check your build definition for
[error] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
[error] A full rebuild may help if 'SparkFunSuite.class' was compiled against an incompatible version of org.apache.spark.internal.
[error] one error found
[error] Compile failed at Apr 9, 2016 9:32:48 AM [1.095s]
~~~

**Comment by mengxr** [comment]
@dbtsai Let's remove the test dependency on spark-core:test and use `FunSuite` with `// scalastyle:ignore funsuite` in mllib-local test, similar to sketch.
